File: lib/src/facts/resolvers/dmi_resolver.cc
#include <facter/facts/resolvers/dmi_resolver.hpp>
#include <facter/facts/collection.hpp>
#include <facter/facts/fact.hpp>
#include <facter/facts/scalar_value.hpp>
#include <facter/facts/map_value.hpp>

using namespace std;

namespace facter { namespace facts { namespace resolvers {

    dmi_resolver::dmi_resolver() :
        resolver(
            "desktop management interface",
            {
                fact::dmi,
                fact::bios_vendor,
                fact::bios_version,
                fact::bios_release_date,
                fact::board_asset_tag,
                fact::board_manufacturer,
                fact::board_product_name,
                fact::board_serial_number,
                fact::chassis_asset_tag,
                fact::manufacturer,
                fact::product_name,
                fact::serial_number,
                fact::uuid,
                fact::chassis_type,
            })
    {
    }

    string dmi_resolver::to_chassis_description(string const& type)
    {
        if (type.empty()) {
            return {};
        }

        static map<string, string> const descriptions = {
            { "1",  "Other" },
            // 2 is Unknown, which we'll output if it's not in the map anyway
            { "3",  "Desktop" },
            { "4",  "Low Profile Desktop" },
            { "5",  "Pizza Box" },
            { "6",  "Mini Tower" },
            { "7",  "Tower" },
            { "8",  "Portable" },
            { "9",  "Laptop" },
            { "10", "Notebook" },
            { "11", "Hand Held" },
            { "12", "Docking Station" },
            { "13", "All in One" },
            { "14", "Sub Notebook" },
            { "15", "Space-Saving" },
            { "16", "Lunch Box" },
            { "17", "Main System Chassis" },
            { "18", "Expansion Chassis" },
            { "19", "SubChassis" },
            { "20", "Bus Expansion Chassis" },
            { "21", "Peripheral Chassis" },
            { "22", "Storage Chassis" },
            { "23", "Rack Mount Chassis" },
            { "24", "Sealed-Case PC" },
        };

        auto it = descriptions.find(type);
        if (it != descriptions.end()) {
            return it->second;
        }
        return "Unknown";
    }

    void dmi_resolver::resolve(collection& facts)
    {
        auto data = collect_data(facts);
        auto dmi = make_value<map_value>();

        auto bios = make_value<map_value>();
        if (!data.bios_vendor.empty()) {
            facts.add(fact::bios_vendor, make_value<string_value>(data.bios_vendor, true));
            bios->add("vendor", make_value<string_value>(move(data.bios_vendor)));
        }
        if (!data.bios_version.empty()) {
            facts.add(fact::bios_version, make_value<string_value>(data.bios_version, true));
            bios->add("version", make_value<string_value>(move(data.bios_version)));
        }
        if (!data.bios_release_date.empty()) {
            facts.add(fact::bios_release_date, make_value<string_value>(data.bios_release_date, true));
            bios->add("release_date", make_value<string_value>(move(data.bios_release_date)));
        }

        auto board = make_value<map_value>();
        if (!data.board_asset_tag.empty()) {
            facts.add(fact::board_asset_tag, make_value<string_value>(data.board_asset_tag, true));
            board->add("asset_tag", make_value<string_value>(move(data.board_asset_tag)));
        }
        if (!data.board_manufacturer.empty()) {
            facts.add(fact::board_manufacturer, make_value<string_value>(data.board_manufacturer, true));
            board->add("manufacturer", make_value<string_value>(move(data.board_manufacturer)));
        }
        if (!data.board_product_name.empty()) {
            facts.add(fact::board_product_name, make_value<string_value>(data.board_product_name, true));
            board->add("product", make_value<string_value>(move(data.board_product_name)));
        }
        if (!data.board_serial_number.empty()) {
            facts.add(fact::board_serial_number, make_value<string_value>(data.board_serial_number, true));
            board->add("serial_number", make_value<string_value>(move(data.board_serial_number)));
        }

        auto product = make_value<map_value>();
        if (!data.product_name.empty()) {
            facts.add(fact::product_name, make_value<string_value>(data.product_name, true));
            product->add("name", make_value<string_value>(move(data.product_name)));
        }
        if (!data.serial_number.empty()) {
            facts.add(fact::serial_number, make_value<string_value>(data.serial_number, true));
            product->add("serial_number", make_value<string_value>(move(data.serial_number)));
        }
        if (!data.uuid.empty()) {
            facts.add(fact::uuid, make_value<string_value>(data.uuid, true));
            product->add("uuid", make_value<string_value>(move(data.uuid)));
        }

        auto chassis = make_value<map_value>();
        if (!data.chassis_asset_tag.empty()) {
            facts.add(fact::chassis_asset_tag, make_value<string_value>(data.chassis_asset_tag, true));
            chassis->add("asset_tag", make_value<string_value>(move(data.chassis_asset_tag)));
        }
        if (!data.chassis_type.empty()) {
            facts.add(fact::chassis_type, make_value<string_value>(data.chassis_type, true));
            chassis->add("type", make_value<string_value>(move(data.chassis_type)));
        }

        if (!data.manufacturer.empty()) {
            facts.add(fact::manufacturer, make_value<string_value>(data.manufacturer, true));
            dmi->add("manufacturer", make_value<string_value>(move(data.manufacturer)));
        }

        if (!bios->empty()) {
            dmi->add("bios", move(bios));
        }

        if (!board->empty()) {
            dmi->add("board", move(board));
        }

        if (!product->empty()) {
            dmi->add("product", move(product));
        }

        if (!chassis->empty()) {
            dmi->add("chassis", move(chassis));
        }

        if (!dmi->empty()) {
            facts.add(fact::dmi, move(dmi));
        }
    }

}}}  // namespace facter::facts::resolvers

File: lib/facter/resolvers/dmi_resolver.rb
# frozen_string_literal: true

module Facter
  module Resolvers
    module Linux
      class DmiBios < BaseResolver
        @log = Facter::Log.new(self)
        @semaphore = Mutex.new
        @fact_list ||= {}

        class << self
          # :bios_vendor
          # :bios_date
          # :bios_version
          # :board_vendor
          # :board_serial
          # :board_name
          # :chassis_asset_tag
          # :chassis_type
          # :sys_vendor
          # :product_serial
          # :product_name
          # :product_uuid

          private

          def post_resolve(fact_name)
            @fact_list.fetch(fact_name) { read_facts(fact_name) }
          end

          def read_facts(fact_name)
            files = %w[bios_date bios_vendor bios_version board_vendor board_name board_serial
                       chassis_asset_tag chassis_type sys_vendor product_name product_serial
                       product_uuid]
            return unless File.directory?('/sys/class/dmi')

            file_content = Util::FileHelper.safe_read("/sys/class/dmi/id/#{fact_name}", nil)
            if files.include?(fact_name.to_s) && file_content
              @fact_list[fact_name] = file_content.strip
              chassis_to_name(@fact_list[fact_name]) if fact_name == :chassis_type

            end
            @fact_list[fact_name]
          end

          def chassis_to_name(chassis_type)
            types = ['Other', nil, 'Desktop', 'Low Profile Desktop', 'Pizza Box', 'Mini Tower', 'Tower',
                     'Portable', 'Laptop', 'Notebook', 'Hand Held', 'Docking Station', 'All in One', 'Sub Notebook',
                     'Space-Saving', 'Lunch Box', 'Main System Chassis', 'Expansion Chassis', 'SubChassis',
                     'Bus Expansion Chassis', 'Peripheral Chassis', 'Storage Chassis', 'Rack Mount Chassis',
                     'Sealed-Case PC', 'Multi-system', 'CompactPCI', 'AdvancedTCA', 'Blade', 'Blade Enclosure',
                     'Tablet', 'Convertible', 'Detachable']
            @fact_list[:chassis_type] = types[chassis_type.to_i - 1]
          end
        end
      end
    end
  end
end
